### PR TITLE
Implement !is operator

### LINKED ISSUE #271: Implement `!is` operator
Follow on from #144 

Specified https://github.com/ballerina-platform/ballerina-spec/issues/752

### PR DESCRIPTION
**Pull request by ushirask**
Implement `!is` operator
Resolves #271

### CODE REVIEW

File: compiler/modules/front.syntax/parseExpr.bal [hunk 1]
```diff
@@ -106,6 +106,19 @@ function parseRelationalExpr(Tokenizer tok) returns Expr|err:Syntax {
         TypeTestExpr typeTest = { td, left: expr, semType: resolveInlineTypeDesc(td) };
         return typeTest;
     }
+    if t == "!" {
+        check tok.advance();
+        Token? t2 = tok.current();
+        if t2 is "is" {
+            tok.setMode(MODE_TYPE_DESC);
```
Comment (jclark) [inline compiler/modules/front.syntax/parseExpr.bal]: Factor common code out.

File: compiler/modules/front.syntax/ast.bal [hunk 2]
```diff
@@ -231,6 +231,12 @@ public type TypeTestExpr record {|
     t:SemType semType;
 |};
 
+public type TypeTestNotExpr record {|
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: What is this?

File: compiler/modules/front.syntax/astString.bal [hunk 3]
```diff
@@ -371,6 +371,18 @@ function exprToWords(Word[] w, Expr expr, boolean wrap = false) {
             w.push(")");
         }
     }
+    else if expr is TypeTestNotExpr {
```
Comment (jclark) [inline compiler/modules/front.syntax/astString.bal]: What is this?

**Review by jclark — COMMENTED** [review]
This isn't ready yet.

File: compiler/modules/front.syntax/ast.bal (hunk 2 above)
Comment (ushirask) [reply compiler/modules/front.syntax/ast.bal]: This to represent expressions with `!is` in the ast and differentiate from `is` expressions

Comment (ushirask) [reply compiler/modules/front.syntax/ast.bal]: should I use the same `TypeTestExpr` for both with a new field to indicate `!is` ?

File: compiler/modules/front.syntax/parseExpr.bal (hunk 1 above)
Comment (ushirask) [reply compiler/modules/front.syntax/parseExpr.bal]: fixed

File: compiler/modules/front.syntax/ast.bal (hunk 2 above)
Comment (jclark) [reply compiler/modules/front.syntax/ast.bal]: You already added the `negated` field to TypeTestExpr for this...

Comment (ushirask) [reply compiler/modules/front.syntax/ast.bal]: I only added negated for the TypeTestInsn in the bir but not the one in the ast. I can do the same to both to keep consistency

Comment (jclark) [reply compiler/modules/front.syntax/ast.bal]: Yes...

File: compiler/modules/front.syntax/parseExpr.bal [hunk 4]
```diff
@@ -99,18 +99,33 @@ function parseRelationalExpr(Tokenizer tok) returns Expr|err:Syntax {
...
+            return typeTest;
+        }
+        return err:syntax("invalid operator");
+    }
     else {
         return expr;
     }
 }
 
+function handleTypeDesc(Tokenizer tok) returns InlineTypeDesc|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front.syntax/parseExpr.bal]: Please choose a name consistent with other names in this file.

File: compiler/modules/front/codeGen.bal [hunk 5]
```diff
@@ -958,22 +958,15 @@ function codeGenExpr(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:Ex
...
-            return codeGenTypeTest(cx, bb, env, td, left, semType);
+            return codeGenTypeTest(cx, bb, env, td, left, semType, false);
+        }
+        //Type not test
+        var { notTd, left, semType } => {
+            return codeGenTypeTest(cx, bb, env, notTd, left, semType, true);
         }
         // Variable reference
         var { varName } => {
-            var v = check lookupVarRef(cx, varName, env);
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal:965]: Why is this needed for `!is`?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 5 above)
Comment (ushirask) [reply compiler/modules/front/codeGen.bal:965]: moved it to avoid method too large error

File: compiler/modules/front.syntax/parseExpr.bal (hunk 4 above)
Comment (ushirask) [reply compiler/modules/front.syntax/parseExpr.bal]: fixed

File: compiler/modules/front.syntax/ast.bal (hunk 2 above)
Comment (ushirask) [reply compiler/modules/front.syntax/ast.bal]: fixed

File: compiler/modules/front/codeGen.bal [hunk 6]
```diff
@@ -1115,24 +1104,39 @@ function codeGenEquality(CodeGenContext cx, bir:BasicBlock bb, Environment env,
     }
 }
 
-function codeGenTypeTest(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:TypeDesc td, s:Expr left, t:SemType semType) returns CodeGenError|ExprEffect {
+function codeGenVarRefExpr(CodeGenContext cx, string name, Environment env, bir:BasicBlock bb) returns CodeGenError|ExprEffect{
+    var v = check lookupVarRef(cx, name, env);
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal:1108]: Indentation looks weird here.

File: compiler/modules/front.syntax/parseExpr.bal [hunk 7]
```diff
@@ -99,18 +99,33 @@ function parseRelationalExpr(Tokenizer tok) returns Expr|err:Syntax {
...
+            return typeTest;
+        }
+        return err:syntax("invalid operator");
+    }
     else {
         return expr;
     }
 }
 
+function parseTestTypeDesc(Tokenizer tok) returns InlineTypeDesc|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front.syntax/parseExpr.bal]: A better way to do this is

- call the function finishTypeTestExpr
- pass in `negated` as a parameter
- return a TypeTestExpr

**Review by jclark — CHANGES_REQUESTED** [review]
Much improved. Nearly there.

File: compiler/modules/front.syntax/parseExpr.bal (hunk 7 above)
Comment (ushirask) [reply compiler/modules/front.syntax/parseExpr.bal]: done
